### ISSUE TITLE
meson incorrectly thinks that gcc-11 doesn't support `-std=c++23`

### ISSUE THREAD

**Issue by bredelings** [mesonbuild/meson#12291]
**Describe the bug**
For a project with `cpp_std=c++23`, meson incorrectly states that 

```
meson.build:1:0: ERROR: Value "c++23" (of type "string") for combo option "C++ language standard to use" is not one of the choices. Possible choices are (as string): "none", "c++98", "c++03", "c++11", "c++14", "c++17", "c++1z", "c++2a", "c++20", "gnu++03", "gnu++11", "gnu++14", "gnu++17", "gnu++1z", "gnu++2a", "gnu++20".
```

**To Reproduce**
Define the project like this:
```
project('bali-phy', ['cpp','c'],
        version: '4.0-beta7-preview',
        default_options : [
          'cpp_std=c++23',
          'warning_level=3',
          'buildtype=release',
          'b_ndebug=if-release'
        ],
        meson_version: '>= 1.1',
        license: 'GPLv2')
```

**Expected behavior**
I have gcc-11 version 11.4.0, and `gcc-11 -std=c++23 test.cc` works fine.  So I expect that meson should not complain.

The problem appears to be that meson has hard-coded 12.2.0 as the first version that can understand c++23.
```
        if version_compare(self.version, '>=12.2.0'):
            cppstd_choices.append('c++23')
```

It seems that all gcc-11 minor versions should allow `-std=c++23`, but I'm not sure this flag has any effect.  If 12.2 is the first supported version because it is the first version where the flag has an effect, then maybe we could add a comment line like
```
        # GCC-11 allows `-std=c++23` but doesn't do anything
        if version_compare(self.version, '>=12.2.0'):
            cppstd_choices.append('c++23')
```

**system parameters**
LInux x86, Debian unstable.
meson version 1.2.1, ninja version 1.11.1

**Comment by dvrogozh** [comment]
Support for c++23 and c++26 are recent additions to meson, see https://github.com/mesonbuild/meson/pull/11986. It was merged in July 31, 2023 as https://github.com/mesonbuild/meson/commit/1cd16a7cc5a19b1d694f66d80f36f6ec4e119dcb. And as far as I see it's not available in any releases yet:
```
$ git tag --contains 1cd16a7cc5a19b1d694f66d80f36f6ec4e119dcb

$ git branch -a --contains 1cd16a7cc5a19b1d694f66d80f36f6ec4e119dcb
* master
  remotes/origin/HEAD -> origin/master
  remotes/origin/master
  remotes/origin/wip/xclaesse/pr-12097
  remotes/origin/wixfix
```

I think you need to either try use meson directly from the git repo or wait for the next release. Checking that meson from latest master works would help to verify that c++23 support is properly works on your side.

**Comment by bredelings** [comment]
I double-checked, and the problematic code IS present in the `master` branch.

```
        if version_compare(self.version, '>=12.2.0'):
            cppstd_choices.append('c++23')
```

A fix just needs to change `>=12.2.0` to `>=11.0`.

There are additionally two cases where the string `12.2.0` shows up in `unittests/linuxliketests.py`.  These should probably be fixed as well.

**Comment by dvrogozh** [comment]
As for the gcc-11, it has **experimental** support for some c++23 features. See https://www.gnu.org/software/gcc/gcc-11/changes.html: "The C++ front end has experimental support for some of the upcoming C++23 draft features with the -std=c++23". Actually, gcc11, gcc12, gcc13 - all of these has only **experimental** c++23 support. Though -std=c++23 switch indeed first appeared in gcc11.

I guess we need to ask an author of this check, @jpakkane, regarding rational behind `12.2.0` limitation:
https://github.com/mesonbuild/meson/blob/32be9bd6de0a4769cdf6d9528496925643739918/mesonbuild/compilers/cpp.py#L429

I wonder what's meson maintainers strategy to allow usage of specific, say c++23, standard? One strategy might be to allow as soon as compiler has a switch - in this case that's gcc11. Other strategy - wait for official support in the compiler, i.e. when standard support will mature from "experimental" - for c++23 today there is no gcc release which "officially" supports it. If standard is allowed somewhere in the middle (which is 12.2.0 limitation for c++23), then that's a risk for the production code to step into compiler bugs which won't be fixed in the compiler version used by the product (for c++23 compiler fixes most likely won't be ported to gcc11-13).

**Comment by tristan957** [comment]
If the compiler has the switch, we should allow it in my opinion.

**Comment by bredelings** [comment]
> I guess we need to ask an author of this check, @jpakkane, regarding rational behind 12.2.0 limitation: 

Yes.

> I wonder what's meson maintainers strategy to allow usage of specific, say c++23, standard?

I was wondering the same thing.  I don't think its workable for `cpp_std=c++23` to require all of c++23 to be implemented, because that would mean that we couldn't use any of the features until all of them were implemented.

**Comment by dvrogozh** [comment]
> I don't think its workable for cpp_std=c++23 to require all of c++23 to be implemented, because that would mean that we couldn't use any of the features until all of them were implemented.

In some cases people might wish to wait for the standard to be declared as officially supported by the compiler (officially - as opposed to experimental). Otherwise you potentially won't have bug fixes support from compiler side for the version of compiler you are using and will be forced to either rewrite your code or switch to later compiler version. This does not mean however, that build system, meson, should follow such a guidance. Project maintainers though should keep this consideration in mind though.

PS: I am typically reading this part of gcc documentation for that: https://gcc.gnu.org/onlinedocs/gcc-12.3.0/gcc/Standards.html#Language-Standards-Supported-by-GCC. For clang I don't know good source of such information unfortunately.

**Comment by eli-schwartz** [comment]
> In some cases people might wish to wait for the standard to be declared as officially supported by the compiler (officially - as opposed to experimental).

Confusingly, the new policy for clang and for GCC seems to be that compiler authors are no longer interested in having a `-std=c++2b` be "the" way to ask for experimental support, so now they're adding the final official public name waaaaay in advance and just not documenting it.

It sounded to me like their reasoning for doing it was because they wanted to avoid the case where you pass such an argument on older gcc and it didn't work, but I thought that was the whole *point* that it shouldn't work.

**Comment by bredelings** [comment]
> It sounded to me like their reasoning for doing it was because they wanted to avoid the case where you pass such an argument on older gcc and it didn't work, but I thought that was the whole point that it shouldn't work.

Based on reading gcc-patches, I think that is maybe not right.  I think that previously people used i.e. `c++2b` not because the standard wasn't fully _implemented_, but because the standard had not yet been _released_.  I think that people were fine with adding `c++20` as soon as the c++20 standard was finalized, even though it still has not been fully implemented in GCC.

Also it may be relevant that people started doing this with c++0x=c++11.  Part of the reason was because they weren't really sure which _year_ the standard was going to get released in.  Since c++11, however, the standards have never missed the three-year release cadence.  So that may explain some part of why people are more comfortable adding `c++23` before 2023.

**Comment by bruchar1** [comment]
Was this issue fixed by #12340 ? If so, it can be closed...

**Comment by bredelings** [comment]
Yes, it should be fixed.
